src/journal-remote/journal-gatewayd.c

```diff
@@ -641,6 +641,66 @@ static int request_handler_fields(
         return MHD_queue_response(connection, MHD_HTTP_OK, response);
 }
 
+static int output_field_name(FILE *f, OutputMode m, const char *d) {
+        if (m == OUTPUT_JSON)
+                return fprintf(f, "{ \"field\" : \"%s\" }\n", d);
+        else
+                return fprintf(f, "%s\n", d);
+}
+
+static int request_handler_field_names(
+                struct MHD_Connection *connection,
+                void *connection_cls) {
+
+        _cleanup_(MHD_destroy_responsep) struct MHD_Response *response = NULL;
+        RequestMeta *m = connection_cls;
+        _cleanup_close_ int fd = -1;
+        _cleanup_fclose_ FILE *tmp = NULL;
+        int r;
+        off_t n;
+        const char *field;
+
+        assert(connection);
+        assert(m);
+
+        r = open_journal(m);
+        if (r < 0)
+                return mhd_respondf(connection, r, MHD_HTTP_INTERNAL_SERVER_ERROR, "Failed to open journal: %m");
+
+        if (request_parse_accept(m, connection) < 0)
+                return mhd_respond(connection, MHD_HTTP_BAD_REQUEST, "Failed to parse Accept header.");
+
+        fd = open_tmpfile_unlinkable("/tmp", O_RDWR|O_CLOEXEC);
+        if (fd < 0)
+                return mhd_respondf(connection, r, MHD_HTTP_INTERNAL_SERVER_ERROR, "Failed to create temporary file: %m");
+
+        tmp = fdopen(fd, "w+");
```

Comment:
Maybe, better to use memstream and dprintf() instead. See src/basic/memstream-util.[ch].